### PR TITLE
Add the function to upload to tgraph

### PR DESCRIPTION
**Pull request by zxcvbnmzsedr**
Upload the image directly to tgraph without proxying through another site.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -38,3 +38,4 @@ CJKwrap==2.2
 typing-extensions==4.10.0
 uvloop==0.19.0; sys_platform!='win32' and sys_platform!='cygwin' and sys_platform!='cli'
 isal==1.6.0; platform_machine=='x86_64' or platform_machine=='AMD64' or platform_machine=='aarch64'
+requests==2.26.0
```
Comment (Rongronggg9) [inline requirements.txt:41]: RSStT is asynchronous, using synchronous HTTP lib will make the bot stuck.
Use aiohttp instead.

File: src/env.py [hunk 2]
```diff
@@ -154,6 +154,7 @@ def __list_parser(var: Optional[str]) -> list[str]:
 API_ID: Final = int(os.environ['API_ID']) if os.environ.get('API_ID') else None
 API_HASH: Final = os.environ.get('API_HASH')
 TOKEN: Final = os.environ.get('TOKEN')
+TELEGRAPH_IMG_UPLOAD: Final = os.environ['TELEGRAPH_IMG_UPLOAD'] if os.environ.get('TELEGRAPH_IMG_UPLOAD') else False
```
Comment (Rongronggg9) [inline src/env.py:157]: Document it in `docs/advanced-settings.md`, `docker-compose.yml.sample` and `.env.smaple`.

File: src/parsing/tgraph.py [hunk 3]
```diff
@@ -299,7 +300,11 @@ async def generate_page(self):
                         if tag.name == 'img' and not attr_content.startswith(env.IMAGES_WESERV_NL):
                             if attr_content.split('.', 1)[1].split('/', 1)[0] == 'sinaimg.cn':
                                 attr_content = env.IMG_RELAY_SERVER + attr_content
-                            attr_content = construct_weserv_url(attr_content)
+                            if env.TELEGRAPH_IMG_UPLOAD:
+                                attr_content = telegraph_file_upload(attr_content)
+                            else:
+                                attr_content = construct_weserv_url(attr_content)
+                            logger.warning(f'Processed img: {attr_content}')
```
Comment (Rongronggg9) [inline src/parsing/tgraph.py:307]: Why warning?

File: src/parsing/tgraph.py [hunk 4]
```diff
@@ -357,3 +362,34 @@ async def telegraph_ify(self):
                     f'Network error ({type(e).__name__}) occurred when creating telegraph page, will retry')
                 return await self.telegraph_ify()
             raise e
+
+
+def telegraph_file_upload(file_url):
```
Comment (Rongronggg9) [inline src/parsing/tgraph.py:367]: What if moving it to `src/web/media.py`?

File: src/parsing/tgraph.py [hunk 5]
```diff
@@ -357,3 +362,34 @@ async def telegraph_ify(self):
...
+    path_to_file -> str, path to a local file
+
+    Return
+    ---------------
+    telegraph_url -> str, url of the file uploaded
+
+    >>>telegraph_file_upload('test_image.jpg')
+    https://telegra.ph/file/16016bafcf4eca0ce3e2b.jpg
+    >>>telegraph_file_upload('untitled.txt')
+    error, txt-file can not be processed
```
Comment (Rongronggg9) [inline src/parsing/tgraph.py:383]: Copied from elsewhere?

**Comment by zxcvbnmzsedr** [comment]
No problem, I didn't consider that much; this way is the quickest and most straightforward solution to my current problem. I will make some adjustments to the code based on your advice.

File: src/parsing/tgraph.py [hunk 6]
```diff
@@ -357,3 +362,34 @@ async def telegraph_ify(self):
...
+    '''
+    response = requests.get(file_url)
+    if response.status_code == 200:
+        # 提取图片内容
+        image_data = response.content
+        url = 'https://telegra.ph/upload'
+        upload_response = requests.post(url, files={'file': ('file', image_data)})
```
Comment (Rongronggg9) [inline src/parsing/tgraph.py:390]: Write a POST wrapper in `src/web/req.py`, which shares most codes with `_get` and `get`, and use it here. Or else proxy and other features will not be applied.

File: src/parsing/tgraph.py [hunk 7]
```diff
@@ -357,3 +362,34 @@ async def telegraph_ify(self):
...
+    telegraph_url -> str, url of the file uploaded
+
+    >>>telegraph_file_upload('test_image.jpg')
+    https://telegra.ph/file/16016bafcf4eca0ce3e2b.jpg
+    >>>telegraph_file_upload('untitled.txt')
+    error, txt-file can not be processed
+    '''
+    response = requests.get(file_url)
+    if response.status_code == 200:
+        # 提取图片内容
```
Comment (Rongronggg9) [inline src/parsing/tgraph.py:387]: Unneeded comment. The code itself is self-describing.

File: src/parsing/tgraph.py [hunk 8]
```diff
@@ -269,7 +272,6 @@ async def generate_page(self):
                 elif tag.name not in TELEGRAPH_ALLOWED_TAGS:
                     tag.replaceWithChildren()  # remove disallowed tags
                     continue
-
```
Comment (Rongronggg9) [inline src/parsing/tgraph.py:272]: Unneeded reformatting.

File: src/parsing/tgraph.py [hunk 9]
```diff
@@ -278,7 +280,6 @@ async def generate_page(self):
                     alt = tag.get('alt')
                     tag.replaceWith(alt) if alt else tag.decompose()  # drop emoticon
                     continue
-
```
Comment (Rongronggg9) [inline src/parsing/tgraph.py:281]: Ditto.

File: src/parsing/tgraph.py [hunk 10]
```diff
@@ -357,3 +362,34 @@ async def telegraph_ify(self):
...
+    Return
+    ---------------
+    telegraph_url -> str, url of the file uploaded
+
+    >>>telegraph_file_upload('test_image.jpg')
+    https://telegra.ph/file/16016bafcf4eca0ce3e2b.jpg
+    >>>telegraph_file_upload('untitled.txt')
+    error, txt-file can not be processed
+    '''
+    response = requests.get(file_url)
```
Comment (Rongronggg9) [inline src/parsing/tgraph.py:385]: Use the `web.get` wrapper (in `src/web/req.py`). Or else proxy and other features will not be applied.

File: requirements.txt (hunk 1 above)
Comment (Rongronggg9) [inline requirements.txt:41]: RSStT is asynchronous, using synchronous HTTP lib will make the bot stuck.
Use the `web` wrapper (in `src/web/req.py`) instead.

File: src/env.py (hunk 2 above)
Comment (Rongronggg9) [inline src/env.py:157]: Document it in `docs/advanced-settings.md`, `docker-compose.yml.sample` and `.env.smaple`.
And use `__bool_parser()` above.

File: src/parsing/tgraph.py (hunk 8 above)
Comment (Rongronggg9) [inline src/parsing/tgraph.py:272]: Unneeded reformatting.

File: src/parsing/tgraph.py (hunk 9 above)
Comment (Rongronggg9) [inline src/parsing/tgraph.py:281]: Ditto.

File: src/parsing/tgraph.py (hunk 3 above)
Comment (Rongronggg9) [inline src/parsing/tgraph.py:307]: Why warning?

File: src/parsing/tgraph.py [hunk 11]
```diff
@@ -357,3 +362,34 @@ async def telegraph_ify(self):
...
+    if response.status_code == 200:
+        # 提取图片内容
+        image_data = response.content
+        url = 'https://telegra.ph/upload'
+        upload_response = requests.post(url, files={'file': ('file', image_data)})
+        telegraph_url = json.loads(upload_response.content)
+        telegraph_url = telegraph_url[0]['src']
+        telegraph_url = f'https://telegra.ph{telegraph_url}'
+        return telegraph_url
+    return construct_weserv_url(file_url)
```
Comment (Rongronggg9) [inline src/parsing/tgraph.py:395]: What about moving it to `src/web/media.py`?

File: src/parsing/tgraph.py (hunk 5 above)
Comment (Rongronggg9) [inline src/parsing/tgraph.py:383]: Copied from else where?

File: src/parsing/tgraph.py [hunk 12]
```diff
@@ -357,3 +362,34 @@ async def telegraph_ify(self):
...
+    ---------------
+    telegraph_url -> str, url of the file uploaded
+
+    >>>telegraph_file_upload('test_image.jpg')
+    https://telegra.ph/file/16016bafcf4eca0ce3e2b.jpg
+    >>>telegraph_file_upload('untitled.txt')
+    error, txt-file can not be processed
+    '''
+    response = requests.get(file_url)
+    if response.status_code == 200:
```
Comment (Rongronggg9) [inline src/parsing/tgraph.py:386]: Use the `web.get` wrapper (in `src/web/req.py`). Or else proxy and other features will not be applied.

File: src/parsing/tgraph.py (hunk 7 above)
Comment (Rongronggg9) [inline src/parsing/tgraph.py:387]: Unneeded comment. The code itself is self-describing.

File: src/parsing/tgraph.py (hunk 6 above)
Comment (Rongronggg9) [inline src/parsing/tgraph.py:390]: Write a POST wrapper in `src/web/req.py`, which shares most codes with `_get` and `get`, and use it here. Or else proxy and other features will not be applied.

File: src/parsing/tgraph.py [hunk 13]
```diff
@@ -357,3 +362,34 @@ async def telegraph_ify(self):
...
+        # 提取图片内容
+        image_data = response.content
+        url = 'https://telegra.ph/upload'
```
Comment (Rongronggg9) [inline src/parsing/tgraph.py:389]: This should be a const variable outside the function.

File: src/parsing/tgraph.py [hunk 14]
```diff
@@ -357,3 +362,34 @@ async def telegraph_ify(self):
...
+        upload_response = requests.post(url, files={'file': ('file', image_data)})
+        telegraph_url = json.loads(upload_response.content)
+        telegraph_url = telegraph_url[0]['src']
+        telegraph_url = f'https://telegra.ph{telegraph_url}'
```
Comment (Rongronggg9) [inline src/parsing/tgraph.py:393]: The domain of Telegraph should be a const variable outside the function.

Comment (Rongronggg9) [inline src/parsing/tgraph.py:393]: All three lines can be inlined into one line.

File: src/parsing/tgraph.py (hunk 12 above)
Comment (Rongronggg9) [inline src/parsing/tgraph.py:386]: Only using a condition is not sufficient. Exception handling (fallback) is needed.

File: src/parsing/tgraph.py [hunk 15]
```diff
@@ -1,7 +1,10 @@
 from __future__ import annotations
+
+import json
```
Comment (Rongronggg9) [inline src/parsing/tgraph.py:3]: Adding `import` should be done on hand. Do not rely on your IDE to do it.
Read the `import` blocks twice to see where to add `import`.

File: src/parsing/tgraph.py [hunk 16]
```diff
@@ -1,7 +1,10 @@
...
 from typing import Union, Optional
 from typing_extensions import Final
 from collections.abc import Awaitable
+import requests
```
Comment (Rongronggg9) [inline src/parsing/tgraph.py:7]: Ditto.

**Review by Rongronggg9 — CHANGES_REQUESTED** [review]
Personally, I do not object to the feature, as long as it is opt-in (it does be in the current revision). But it is more like a "dirty hack" as far as I can see now, so more polishing is needed before it can be merged.

**Comment by Rongronggg9** [comment]
Oh, I thought of a much more elegant solution for your use case just now.
Upload media to Telegraph via https://github.com/Rongronggg9/rsstt-img-relay so that it can be enabled for all users (opt-out).

For now, I am going to close your PR. I will implement the feature I mentioned above soon in the following weeks - you don't need to do anything further but just looking forward to it!

**Comment by Rongronggg9** [comment]
> I will implement the feature I mentioned above soon in the following weeks

Such a feature has been implemented in https://github.com/Rongronggg9/rsstt-img-relay/commit/750210443882d2c35e5f4f92f50d0bdd748af4e2. E.g., `/upload_graph/https://example.com/example.jpg`.

The next step is to implement the corresponding feature in RSStT to take advantage of it.

**Comment by zxcvbnmzsedr** [comment]
I found that tgraph has certain restrictions on uploaded media files, such as requiring the dimensions to be within 5000*4000, the file size to be less than 10MB, and not accepting webp images. Uploading images that do not meet these requirements will result in an error message "Image dimensions invalid". Would handling this in JS be very cumbersome?
@Rongronggg9

**Comment by Rongronggg9** [comment]
> Would handling this in JS be very cumbersome?

No need to actually "handle" it. All we need to do is to invoke images.weserv.nl on demand.

**Comment by Rongronggg9** [comment]
> All we need to do is to invoke [images.weserv.nl](https://images.weserv.nl/) on demand.

Done in https://github.com/Rongronggg9/rsstt-img-relay/commit/711597605b9797ade60c7cadc67e6a65ced3c157

**Comment by Rongronggg9** [comment]
> file size to be less than 10MB

The limit should be 5MiB.

> Image dimensions invalid

Reconsider the case, we'd better let it fail in order not to resize long images (words within them will become invisible after resizing).
If an image is not too long, it is very unlikely to hit the dimension limit before hitting the size limit. In other words, we should filter out the "Image dimensions invalid" error message, preventing it from causing images to be resized.

**Comment by Rongronggg9** [comment]
All done 🎉

**Pull request closed without merging**